### PR TITLE
Onboarding Wizard

### PR DESCRIPTION
**Pull request by EDsCODE**
## Changes
- adding onboarding wizard upon initial sign in to guide users through the library/snippet installation

## Checklist
- [ ] All querysets/queries filter by Team (if applicable)
- [ ] Backend tests (if applicable)
- [ ] Cypress E2E tests (if applicable)

### CODE REVIEW

**Comment by EDsCODE** [comment]
Low Quality GIF

onboarding-wizard

**Comment by EDsCODE** [comment]
RE: vertical vs horizontal @fuziontech def can make it vertical! I was going to include back buttons to allow users to return to previous step anyway. not sure if you have preference on direction given the back button

**Comment by fuziontech** [comment]
I think a normal back button (or just scroll if possible) would be awesome

**Comment by timgl** [comment]
This is so cool!!

Wlil cast a vote towards keeping it horizontal as that sort of indicates 'steps' but easy either way.

EDIT: I do think we should push people towards installing the posthog-js snippet first, as it's much easier and gives you access to feature flags and the toolbar. Right now people might think "my app is a python app" but it'd be much harder to get started.

The flow could be "It's a web app" -> install snippet -> "do you want to capture any backend events"

**Comment by mariusandra** [comment]
Very cool! Re: flow, both options (back button and scrolling) would work for me. I have a slight preference for scrolling, as then everything would reveal and scroll downwards... and you can quickly scroll back up to e.g. try the instructions for a different stack.

Something like typeform perhaps:
https://www.typeform.com/templates/t/customer-onboarding-feedback-survey-template/

That said, having just a back button is definitely good enough and it'll already spark most of the joy that can be sparked :D.

**Comment by EDsCODE** [comment]
Ended up removing the animations

File: frontend/src/layout/onboarding/FrameworkInstructions/go_snippet.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+import React from 'react'
+import Snippet from './snippet'
+
```
Comment (EDsCODE) [inline frontend/src/layout/onboarding/FrameworkInstructions/go_snippet.js:3]: This format of splitting snippets is repeated across all frameworks so we could modify this to be way more reusable. However, I manually separated each one for now incase there's some extra information that's needed depending on the framework

**Comment by macobo** [comment]
Fly-by review, feel free to ignore.

Reading the code got me thinking about the `installed_snippet` column.

1. The column is a slight misnomer. It's more along the lines "user passed X tutorial" rather than them _actually_ having installed the tutorial.
2. I'm not sure putting it in the user level is the correct place. Why not team? :)
3. It has small consistency concerns: I.e. what if someone wipes all events and then a user gets added?

This said, all of these are relatively small concerns and outweighed by the functionality being added.

Maybe instead of adding a column, this is a interesting usecase for feature flags? :) I.e. raise a flag for a specific user when they complete this tutorial.

File: frontend/src/scenes/onboarding/onboardingWizard.js [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+        </div>
+    )
+}
+
+function VerificationPanel({ reverse }) {
+    const { loadUser, userUpdateRequest } = useActions(userLogic)
+    const { user } = useValues(userLogic)
+
+    useInterval(() => {
+        !user.has_events && loadUser()
```
Comment (macobo) [inline frontend/src/scenes/onboarding/onboardingWizard.js:218]: This is a really neat way to do the background refresh. :+1:

File: frontend/src/scenes/App.js [hunk 3]
```diff
@@ -10,6 +10,7 @@ import { ToastContainer, Slide } from 'react-toastify'
 import { Sidebar } from '~/layout/Sidebar'
 import { TopContent } from '~/layout/TopContent'
 import { SendEventsOverlay } from '~/layout/SendEventsOverlay'
+import { OnboardingWizard } from '~/scenes/onboarding/onboardingWizard'
```
Comment (macobo) [inline frontend/src/scenes/App.js:13]: Q: Would it make sense to load this + its dependencies in separate chunks as well?

Not a huge issue, but from what I can tell from the code this does make the JS payload larger on every page with the instructions for every platform

**Review by macobo — COMMENTED** [review]
This is neat! Left a few comments/questions.

File: frontend/src/scenes/App.js (hunk 3 above)
Comment (EDsCODE) [reply frontend/src/scenes/App.js:13]: Yes, good call!

**Comment by EDsCODE** [comment]
@macobo Thanks for the review! 

RE: `installed_snippet`, I agree. I was rereading a ticket elsewhere and it reminded me this would be a bit of a strange label on what we're looking for. WiIl make the changes and the lazy loading recommended above

**Comment by EDsCODE** [comment]
@macobo made the changes. Feel free to take another look!

**Comment by macobo** [comment]
@EDsCODE Very cool! LGTM, though it seems like some e2e tests are failing?

**Comment by EDsCODE** [comment]
@macobo mm there's a bit of inconsistency on our cypress tests currently. i'm fairly certain the failing dashboard test is unrelated to this PR. will investigate

**Comment by fuziontech** [comment]
Looks like the first page of the onboarding flow
https://dashboard.cypress.io/projects/twojfp/runs/135/specs

Can't find some element?
image

```
  1) Dashboards
       "before each" hook for "Dashboards loaded":
     AssertionError: Timed out retrying: Expected to find element: `[data-attr=menu-item-dashboards]`, but never found it.
```

**Comment by EDsCODE** [comment]
@fuziontech the test itself is a bit inconsistent. just ran it again off a blank commit and it succeeded :/

**Review by fuziontech — APPROVED** [review]
Hey! The code all looks good and I just tested this out locally.

A few design feedback points:
- The autocapture vs custom capture buttons. It wasn't totally clear to me that those were buttons I could click. I would probably  change the styling a bit to emphasize that they are tabs.
- No code styling on the custom capture tab. On the Autocapture tab the styling looks great for the snippet, but on the Custom Capture page the code is not stylized which is a bit inconsistent.
- The only bit of code that look stylized is the autocapture snippet.

Otherwise the flow looks awesome. This is huge progress. I can't wait to create a funnel for the wizard ;)

File: cypress/support/index.js [hunk 4]
```diff
@@ -36,6 +36,18 @@ beforeEach(() => {
 
             cy.get('.btn').click()
         }
+        cy.wait(2000)
```
Comment (EDsCODE) [inline cypress/support/index.js:39]: wait statement going in here. the conditional testing below needs to get accurate hit on the elements in body. without the wait, the condition would be checked instantly after signup/login which wouldn't give enough time for onboarding wizard to appear

**Comment by timgl** [comment]
@EDsCODE Just coming back to this, I think the flow is nice, however I'd be very surprised if this has increased the number of people that manage to send events. From what I've seen, there's now absolutely no way of seeing the normal JS snippet? However, that is far and away the easiest way to get started for most people.

I'll repeat the point I made in the comment above: I think we should push people first and foremost to use the JS snippet (not the module!), and only if they really don't want that give them the other libraries.

**Comment by EDsCODE** [comment]
The normal JS snippet is the first thing thats shown on any of instructions if youre using Web. You have to click a show more button if you want to use the dedicated library.

The flow is meant to prepare the instructions so everything is in 1 ppace and we can get more info. 

Select platform -> select framework -> user gets shown autocapture snippet if web had been selected or user can choose to integrate module -> polling for events

**Comment by timgl** [comment]
ahhh mea culpa, you're absolutely right. Teaches me for not properly clicking around. Sorry about that!

**Pull request merged**